mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -268,8 +268,26 @@ class BlockMatrix @Since("1.3.0") (
   def toIndexedRowMatrix(): IndexedRowMatrix = {
     require(numCols() < Int.MaxValue, "The number of columns must be within the integer range. " +
       s"numCols: ${numCols()}")
-    // TODO: This implementation may be optimized
-    toCoordinateMatrix().toIndexedRowMatrix()
+
+    val rows = blocks.map(block => (block._1._1, (block._1._2, block._2)))
+      .groupByKey()
```

Comment:
This assumes that a partition can hold an entire block row, which is not always valid. I would suggest the following:

* for each block, break the matrix block into rows and then emit (rowIdx, row). You can map the matrix block to a breeze matrix, and then call rows.
* call groupByKey and then concat breeze vectors

This could be a more scalable implementation.